### PR TITLE
5.0 entity generalization

### PR DESCRIPTION
**Pull request by VISHNUDAS-tunerlabs**

### CODE REVIEW

File: controllers/v1/observationsController.js [hunk 1]
```diff
@@ -2160,4 +2247,53 @@ module.exports = class Observations extends Abstract {
...
+        matchingData.push(entity)
+      } else {
+        mismatchEntities.push(entity.id)
+      }
+    });
+
+    if( mismatchEntities.length > 0 ){
+      await getSubEntitiesBasedOnEntityType(mismatchEntities, entityType, matchingData)
+    } 
+    let uniqueEntities = [];
```
Comment (aks30) [inline controllers/v1/observationsController.js:2293]: @VISHNUDAS-tunerlabs as discussed this is not required. Also please make variable name changes.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 22 Aug.

File: controllers/v1/observationSubmissionsController.js [hunk 2]
```diff
@@ -99,45 +100,50 @@ module.exports = class ObservationSubmissions extends Abstract {
...
+        
+        if ( !entitiesDocument.success ) {
+            return resolve({ 
+                status: httpStatusCode.bad_request.status, 
+                message: messageConstants.apiResponses.ENTITY_NOT_FOUND
+            });
         }
-
+        
+        let entityResult = entitiesDocument.data;
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply controllers/v1/observationSubmissionsController.js:141]: @aks30 returning object argument is not needed. only using at two places and they require object in return.

File: module/entities/helper.js [hunk 3]
```diff
@@ -1883,31 +1902,73 @@ module.exports = class EntitiesHelper {
    */
 
   static observationSearchEntitiesResponse(entities,observationEntityIds) {
-
+
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/entities/helper.js:1905]: @aks30 entity data contain id not _id, it does not contain districtName also

[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/entities/helper.js:1905]: @aks30 we have id instead _id. districtName not available in learners search

File: module/entities/helper.js [hunk 4]
```diff
@@ -926,106 +928,104 @@ module.exports = class EntitiesHelper {
...
+ static validateEntities(entityIds, entityType) {
+    return new Promise(async (resolve, reject) => {
+        try {
+             //set request body for learners API
+            let locationIds = [];
+            let locationCodes = [];
+            let entityInformations = [];
+            let validEntityIds = [];
+            entityIds.forEach(entity=>{
+                if (gen.utils.checkIfValidUUID(entity)) {
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/entities/helper.js:980]: @aks30 old code -old code we were checking with entityId. But now we need to check uuid and code (for school)

File: controllers/v1/observationSubmissionsController.js [hunk 5]
```diff
@@ -194,7 +200,7 @@ module.exports = class ObservationSubmissions extends Abstract {
         
 
         let submissionDocument = {
-          entityId: entityDocument._id,
+          entityId: entityDocument.id,
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply controllers/v1/observationSubmissionsController.js:197]: @aks30 change made.

File: module/observations/helper.js [hunk 6]
```diff
@@ -1756,49 +1775,76 @@ module.exports = class ObservationsHelper {
   static listEntities( observationId ) {
     return new Promise(async (resolve, reject) => {
         try {
-            
+
             let observationDocument = await this.observationDocuments({
                 _id : observationId
             },["entities","entityType"]);
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/observations/helper.js:1781]: @aks30 ok

File: module/observations/helper.js [hunk 7]
```diff
@@ -1756,49 +1775,76 @@ module.exports = class ObservationsHelper {
...
                         observationId : observationId,
                         entityId : currentEntities._id
                     });
-
+                    
                     let entity = {
-                        _id : currentEntities._id,
-                        externalId : currentEntities.metaInformation.externalId,
-                        name : currentEntities.metaInformation.name,
+                        _id : currentEntities.id,
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/observations/helper.js:1827]: @aks30 added.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 24 Aug

**Pull request merged**